### DISCUSSION TITLE
Python 3.5 Support

### DISCUSSION THREAD

**Discussion by denewiler** [sscpac/statick#290] [General]
## Python 3.5 Support

We currently support Python 3.5 because that is the default Python version in Ubuntu 16.04.

https://packages.ubuntu.com/xenial/python/python3

Ubuntu 16.04 will reach its end-of-life in April 2021.

https://endoflife.software/operating-systems/linux/ubuntu

Python 3.5 already reached its end-of-life in September 2020.

https://www.python.org/dev/peps/pep-0478/

Based on all of this we plan to drop support for Python 3.5 after the Ubuntu 16.04 end-of-life date of April 2021.

### Task List

When Python 3.5 is no longer supported by Python there are a few tasks we can work on.

- [x] Install and run black for all Python versions in CI jobs. Note that black is not available in Python 3.5.
- [x] Update type hints syntax from comment style to inline style.
  Change from
  ```python
  my_variable = 1  # type: int
  ```
  to
  ```python
  my_variable: int = 1
  ```
- [ ] Switching to use of pathlib is also recommended. It is safer and results in cleaner code.
      https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep519
- [x] Switch to f-strings for all print statements. Details in [PEP 498](https://www.python.org/dev/peps/pep-0498/). F-strings are considered more readable and safer to use due to being less error-prone. Do not use f-strings with the `logging` module, as it can introduce a lot of overhead: https://docs.python.org/3/howto/logging.html#optimization.
- [x] Add plugin for isort. The isort tool only runs on Python 3.6+ (at least the part about respecting the `noqa` flag). It is difficult to provide a sample file to trigger isort issues for unit tests without triggering isort for tox. Note that the branch exists at https://github.com/tdenewiler/statick/tree/isort.
- [x] Remove Actions support for Ubuntu 16.04 and Python 3.5.
- [x] Remove deprecated plugins. Those are the `catkin` discovery plugin and the `print_json` reporting plugin. The dependency on the `deprecated` module can then be removed. That allows the `--allow-untyped-decorators` flag to be removed from mypy in Actions.

**Comment by denewiler** [comment]
Switching to use of `pathlib` is also recommended. It is safer and results in cleaner code.

https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-pep519

**Comment by denewiler** [comment]
List of updates for Python 3.6: https://docs.python.org/3/whatsnew/3.6.html

Has more information about type hints for variables.

https://docs.python.org/3/whatsnew/3.6.html#pep-526-syntax-for-variable-annotations

**Comment by tdenewiler** [comment]
We should switch to f-strings for all print statements. Details in [PEP 498](https://www.python.org/dev/peps/pep-0498/). F-strings are considered more readable and safer to use due to being less error-prone.

**Reply by tdenewiler** [reply]
We have very few print statements after switching to the `logging` module. It is not recommended to use f-strings with the `logging` module due to the potential for lots of processing overhead: https://docs.python.org/3/howto/logging.html#optimization. Switching to f-strings will likely only take place in some unit tests, in the print to console reporting plugin, and the json reporting plugin.

**Comment by tdenewiler** [comment]
The ROS Kinetic end-of-life date tracks Ubuntu Xenial: April 2021.

http://wiki.ros.org/Distributions

**Reply by tdenewiler** [reply]
April 26, 2021 has been announced as the date for the final ROS Kinetic sync of packages to the public apt repository.

https://discourse.ros.org/t/preparing-for-final-kinetic-sync/19805
